myriad-scheduler/src/main/java/com/ebay/myriad/scheduler/OfferLifecycleManager.java

```diff
@@ -0,0 +1,86 @@
+package com.ebay.myriad.scheduler;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.inject.Inject;
+
+import org.apache.mesos.Protos;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Manages the Mesos offers tracked by Myriad.
+ */
+public class OfferLifecycleManager {
+  private static final Logger LOGGER = LoggerFactory.getLogger(
+      OfferLifecycleManager.class);
+
+  private Map<String, OfferFeed> offerFeedMap;
+
+  /**
+   * !!! Not thread safe !!!
+   */
+  private final Map<String, ConsumedOffer> consumedOfferMap;
+
+  private final NodeStore nodeStore;
+  private final MyriadDriver myriadDriver;
+
+  @Inject
+  public OfferLifecycleManager(NodeStore nodeStore,
+      MyriadDriver myriadDriver) {
+
+    this.offerFeedMap = new ConcurrentHashMap<>(200, 0.75f, 50);
+    this.consumedOfferMap = new HashMap<>(200, 0.75f);
+    this.nodeStore = nodeStore;
+    this.myriadDriver = myriadDriver;
+  }
+
+  public OfferFeed getOfferFeed(String hostname) {
+    return offerFeedMap.get(hostname);
+  }
+
+  public void declineOffer(Protos.Offer offer) {
+    myriadDriver.getDriver().declineOffer(offer.getId());
+    LOGGER.debug("Declined offer {}", offer.getId());
+  }
+
+  public void addOffers(List<Protos.Offer> offers) {
+    for (Protos.Offer offer : offers) {
+      String hostname = offer.getHostname();
+      Node node = nodeStore.getNode(hostname);
+      if (node != null) {
+        OfferFeed feed = offerFeedMap.get(hostname);
+        if (feed == null) {
+          feed = new OfferFeed();
+          offerFeedMap.put(hostname, feed);
+        }
+        feed.add(offer);
+
+        node.setSlaveId(offer.getSlaveId());
+
+        LOGGER.debug("addResourceOffers: caching offer for host {}, offer id {}",
+            hostname, offer.getId().getValue());
+      } else {
+        myriadDriver.getDriver().declineOffer(offer.getId());
+        LOGGER.debug("Declined offer for unregistered host {}", hostname);
+      }
+    }
```

Comment:
The offers are not cached for a long time. They are cached until the next scheduling cycle within RM, which, by default, is triggered by a NodeManager's HB. When a NM heartbeats with an RM, RM figures out of that NM needs to be allocated with new containers. 

If an NM gets new containers during HB, then Myriad uses the "launchTasks" API to tell mesos that the offer is being used to accommodate for the YARN's containers. Only the resources required for running the new containers will be utilized from the offer. For e.g. if the offer has 10G and 10CPUs and RM allocates 4 containers each with 1G and 1CPU, the "launchTasks" API tells mesos that there will be 4 mesos tasks with 1G and 1CPU to act as "placeholders" for the YARN containers. The remaining 6G and 6CPU from the offer is given back to mesos for other frameworks to use.

OTOH, if the scheduling didn't result in new containers to be allocated, then Myriad simply rejects the offer so that mesos can claim back the resources and offer them to other frameworks.